pkg/storage/unified/resource/search_server_distributor.go

```diff
@@ -114,6 +118,98 @@ func (ds *distributorServer) GetStats(ctx context.Context, r *resourcepb.Resourc
 	return client.GetStats(ctx, r)
 }
 
+func (ds *distributorServer) RebuildIndexes(ctx context.Context, r *resourcepb.RebuildIndexesRequest) (*resourcepb.RebuildIndexesResponse, error) {
+	ctx, span := ds.tracing.Start(ctx, "distributor.RebuildIndexes")
+	defer span.End()
+
+	// distribute the request to all search pods to minimize risk of stale index
+	// it will not rebuild on those which don't have the index open
+	rs, err := ds.ring.GetAllHealthy(searchRingRead)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all healthy instances from the ring")
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		md = make(metadata.MD)
+	}
+
+	var wg sync.WaitGroup
+	var totalRebuildCount atomic.Int64
+	detailsCh := make(chan string, len(rs.Instances))
+	errorCh := make(chan error, len(rs.Instances))
+
+	for _, inst := range rs.Instances {
+		wg.Add(1)
+		go func(inst ring.InstanceDesc) {
+			defer wg.Done()
+
+			client, err := ds.clientPool.GetClientForInstance(inst)
+			if err != nil {
+				ds.log.Error("failed to get client for instance", "instance", inst.Id, "err", err)
+				errorCh <- fmt.Errorf("instance %s: failed to get client, %w", inst.Id, err)
+				return
+			}
+
+			// context for each req
+			instanceCtx := userutils.InjectOrgID(
+				metadata.NewOutgoingContext(ctx, md),
+				r.Namespace,
```

Comment:
Distributor should check that `r.Namespace` and namespaces in keys are the same and return error if not, to detect problems early.